applications/DEMApplication/custom_strategies/strategies/explicit_solver_continuum.cpp

```diff
@@ -399,65 +399,125 @@ namespace Kratos {
         double standard_dev = 0;
-        double out_coordination_number = ComputeCoordinationNumber(standard_dev);
+        double current_coordination_number = ComputeCoordinationNumber(standard_dev);
         int iteration = 0;
         int maxiteration = 100;
         double& added_search_distance = r_process_info[SEARCH_RADIUS_INCREMENT];
+        const bool local_coordination_option = r_process_info[LOCAL_COORDINATION_NUMBER_OPTION];
+        const bool global_coordination_option = r_process_info[GLOBAL_COORDINATION_NUMBER_OPTION];
         added_search_distance = 0.0;
         double amplification = 1.0;
 
         if(r_model_part.GetCommunicator().MyPID() == 0) {
-            KRATOS_WARNING("DEM") << "Setting up Coordination Number by increasing or decreasing the search radius... " << std::endl;
-        }
+            KRATOS_INFO("DEM") << "Setting up Coordination Number (input = "<<desired_coordination_number<<") by increasing or decreasing the search radius. ";
+            if(local_coordination_option) KRATOS_INFO("") << "Local extension activated. ";
+            if(global_coordination_option) KRATOS_INFO("") << "Global extension activated. ";
+            KRATOS_INFO("DEM") << std::endl;
+        }
+
+        if (desired_coordination_number <= 0.0) {
+            KRATOS_THROW_ERROR(std::runtime_error, "The specified Coordination Number is less or equal to zero, N.C. = ", desired_coordination_number)
+        }
+
+        //STAGE 1, Local Coordination Number
+        double tolerance = 0.5;
+        double max_factor_between_iterations = 1.02;
+        double relative_error = 1000;
+        if(local_coordination_option) {
+            while (std::abs(relative_error) > tolerance) {
+                if (iteration >= maxiteration) break;
+                iteration++;
+                if (current_coordination_number == 0.0) {
+                    KRATOS_WARNING("DEM") << "Coordination Number method not supported in this case" << "\n" << std::endl;
+                    KRATOS_THROW_ERROR(std::runtime_error, "The specified tangency method is not supported for this problem, please use absolute value instead", " ")
+                    break;
+                }
 
-        if (in_coordination_number <= 0.0) {
-            KRATOS_THROW_ERROR(std::runtime_error, "The specified Coordination Number is less or equal to zero, N.C. = ", in_coordination_number)
-        }
-        while (fabs(out_coordination_number / in_coordination_number - 1.0) > 1e-3) {
-            if (iteration >= maxiteration) break;
-            iteration++;
-            if (out_coordination_number == 0.0) {
-                KRATOS_WARNING("DEM") << "Coordination Number method not supported in this case" << "\n" << std::endl;
-                KRATOS_THROW_ERROR(std::runtime_error, "The specified tangency method is not supported for this problem, please use absolute value instead", " ")
-                break;
-            }
+                std::vector<double> total_error;
```

Comment:
I think this would be much easier if you use a reduction

e.g. like this:
https://github.com/KratosMultiphysics/Kratos/blob/6c6c37a10415592ca0cf2d47c315c64b25b9a9f6/applications/RANSApplication/custom_processes/auxiliary_processes/rans_nut_y_plus_wall_function_sensitivities_process.cpp#L71-L81

Reply:
Lol! Actually, I have had many bad experiences with openmp reductions. Usually with signed ints, getting totally unexpected results. @ipouplana  can confirm. I am so afraid of using them that I'd rather not....

Reply:
oh really? Good to know
FYI @sunethwarna